draft-ietf-quic-transport.md

```diff
@@ -1471,57 +1502,56 @@ migration, see {{migrate-validate}}.
+token's authenticity can be checked (see {{token-integrity}}), if the client is able
 to return that token, it proves to the server that it received the token.
 
-During the processing of the cryptographic handshake messages from a client, TLS
-will request that QUIC make a decision about whether to proceed based on the
-information it has.  TLS will provide QUIC with any token that was provided by
-the client.  For an initial packet, QUIC can decide to abort the connection,
-allow it to proceed, or request address validation.
-
-If QUIC decides to request address validation, it provides the cryptographic
-handshake with a token.  The contents of this token are consumed by the server
-that generates the token, so there is no need for a single well-defined format.
-A token could include information about the claimed client address (IP and
-port), a timestamp, and any other supplementary information the server will need
-to validate the token in the future.
-
-The cryptographic handshake is responsible for enacting validation by sending
-the address validation token to the client.  A legitimate client will include a
-copy of the token when it attempts to continue the handshake.  The cryptographic
-handshake extracts the token then asks QUIC a second time whether the token is
-acceptable.  In response, QUIC can either abort the connection or permit it to
-proceed.
+If QUIC decides to request address validation for the current
+handshake, it encodes the token in a Retry packet. The contents of
+this token are consumed by the server that generates the token, so
+there is no need for a single well-defined format.  A token could
+include information about the claimed client address (IP and port), a
+timestamp, and any other supplementary information the server will
+need to validate the token in the future.
+
+The Retry packet is sent to the client and then a legitimate client will
+respond with a token in the Initial packet's header when it attempts to continue
+the connection attempt.  In response to receiving the token, QUIC can either
+abort the connection or permit it to proceed.
 
 A connection MAY be accepted without address validation - or with only limited
 validation - but a server SHOULD limit the data it sends toward an unvalidated
 address.  Successful completion of the cryptographic handshake implicitly
 provides proof that the client has received packets from the server.
 
+The client should allow for additional Retry packets being sent in response to
+Initial packets sent containing a token. There are several situations in which
+the server might not be able to use the previously generated token to validate
+the client's address and must send a new Retry. A reasonable limit to the number
+of tries the client allows for, before giving up, is 3. That is, the should
+echo the address validation token from a new Retry packet up to 3 times. After
+that, the client may give up on the connection attempt.
```

Comment:
What is the motivation for this? 3 RTTs is a rather long time to stop a connection attempt.
Why would the server not be able to validate the client's address? This applies only when it crashed and lost the key to generate the token, right? In that case, two retries should be sufficient though.

Reply:
I believe that @nibanks's theory is that there might be multiple layers of DoS devices.

Reply:
@ekr that's my understanding as well. Perhaps @nibanks can provide more details?

Reply:
This is ugly.  Do we currently allow this in QUIC?  I know only the first RETRY can change the connection ID at the moment, so I'd state that.

Reply:
At the very least, I want to support the setup where there is a QUIC server with a completely independent DoS protection device in front of it. The QUIC server should not be required to know anything about the DoS protection device. If we only allow one Retry, per connection attempt, what happens if the real QUIC server ends up trying to do a Retry (because they have a large certificate for instance)?

In the meetings there were other scenarios discussed. For instance, related to dynamic load balancing that for some reason changed between each Initial packet received, causing additional Retries to be sent.